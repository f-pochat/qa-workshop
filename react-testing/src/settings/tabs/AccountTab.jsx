import React from 'react'
import { Button } from "@/components/ui/button"
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/components/ui/card"
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form"
import { Input } from "@/components/ui/input"
import { zodResolver } from "@hookform/resolvers/zod"
import { useForm } from "react-hook-form"
import { z } from "zod"

const formSchema = z.object({
	name: z.string().min(2, "Name too short").max(50, "Name too long"),
	username: z.string().min(2, "Username too short").max(50, "Username too long"),
})

export const AccountTab = () => {
	const form = useForm({
		resolver: zodResolver(formSchema),
		defaultValues: {
			name: 'Cianca',
			username: 'dinos1337'
		}
	})

	const onSubmit = (values) => {
		// some kind of API call should happen here.
		console.log({ values })
	}

	return (
		<Card>
			<CardHeader>
				<CardTitle>Account</CardTitle>
				<CardDescription>
					Make changes to your account here. Click submit when you are done.
				</CardDescription>
			</CardHeader>
			<CardContent className="space-y-2">
				<Form {...form}>
					<form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
						<FormField
							control={form.control}
							name="name"
							render={({ field }) => (
								<FormItem>
									<FormLabel>Name</FormLabel>
									<FormControl>
										<Input placeholder="name" {...field} />
									</FormControl>
									<FormDescription>
										This is your public display name.
									</FormDescription>
									<FormMessage />
								</FormItem>
							)}
						/>
						<FormField
							control={form.control}
							name="username"
							render={({ field }) => (
								<FormItem>
									<FormLabel>Username</FormLabel>
									<FormControl>
										<Input placeholder="username" {...field} />
									</FormControl>
									<FormDescription>
										This will be used for logging in.
									</FormDescription>
									<FormMessage />
								</FormItem>
							)}
						/>
						<Button type="submit">Submit</Button>
					</form>
				</Form>
			</CardContent>
		</Card>
	)
}
