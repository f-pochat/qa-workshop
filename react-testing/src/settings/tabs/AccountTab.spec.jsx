import React from 'react'
import { render, screen, fireEvent, waitFor } from '@/test'
import { describe, expect, test, vi } from 'vitest'
import { AccountTab } from './AccountTab'
import userEvent from '@testing-library/user-event'
import { useForm } from 'react-hook-form'

describe('AccountTab', () => {
	test('should match snapshot', () => {
		const { container } = render(<AccountTab />)
		expect(container.firstChild).toMatchSnapshot()
	})

	test('submitting calls console.log with expected arguments', async () => {
		const consoleSpy = vi.spyOn(console, 'log')
		render(<AccountTab />)
		fireEvent.click(screen.getByRole('button', { name: 'Submit' }))

		// https://react-hook-form.com/advanced-usage#TestingForm
		// waitFor is needed because react-hook-form internally uses asynchronous validation handlers.
		// in order to compute the formState, it has to initially validate the form,
		// which is done asynchronously, resulting in another render.
		await waitFor(() => {
			expect(consoleSpy).toHaveBeenCalledWith({
				values: {
					name: 'Cianca',
					username: 'dinos1337'
				}
			})
		})
	});

	test('Submitting the form with valid data', async () => {
		const consoleSpy = vi.spyOn(console, 'log');
		const { getByPlaceholderText, getByRole } = render(<AccountTab />)

		await userEvent.clear(getByPlaceholderText('name'))
		await userEvent.clear(getByPlaceholderText('username'))

		await userEvent.type(getByPlaceholderText('name'), 'Nicou')
		await userEvent.type(getByPlaceholderText('username'), 'ulmete')

		fireEvent.click(getByRole('button', { name: 'Submit' }))

		// Expectations
		await waitFor(() => {
			expect(consoleSpy).toHaveBeenCalledWith({
				values: {
					name: 'Nicou',
					username: 'ulmete'
				}
			})
		})
	});

	test('Submitting the form with invalid data', async () => {
		const {getByPlaceholderText, getByRole, getByText} = render(<AccountTab />)
		
		await userEvent.clear(getByPlaceholderText('name'))
		await userEvent.clear(getByPlaceholderText('username'));

		await userEvent.type(getByPlaceholderText('username'), Array(51).fill("a").join(""));
	
		fireEvent.click(getByRole('button', { name: 'Submit' }))
	
		await waitFor(() => {
		  expect(getByText("Name too short")).toBeInTheDocument()
		  expect(getByText("Username too long")).toBeInTheDocument()
		})
	  })
})