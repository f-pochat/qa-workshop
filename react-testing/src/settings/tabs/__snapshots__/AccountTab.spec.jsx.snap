// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccountTab > should match snapshot 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm"
>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="text-2xl font-semibold leading-none tracking-tight"
    >
      Account
    </h3>
    <p
      class="text-sm text-muted-foreground"
    >
      Make changes to your account here. Click submit when you are done.
    </p>
  </div>
  <div
    class="p-6 pt-0 space-y-2"
  >
    <form
      class="space-y-8"
    >
      <div
        class="space-y-2"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          for=":r0:-form-item"
        >
          Name
        </label>
        <input
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
          id=":r0:-form-item"
          name="name"
          placeholder="name"
          value="Cianca"
        />
        <p
          class="text-sm text-muted-foreground"
          id=":r0:-form-item-description"
        >
          This is your public display name.
        </p>
      </div>
      <div
        class="space-y-2"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          for=":r1:-form-item"
        >
          Username
        </label>
        <input
          aria-describedby=":r1:-form-item-description"
          aria-invalid="false"
          class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
          id=":r1:-form-item"
          name="username"
          placeholder="username"
          value="dinos1337"
        />
        <p
          class="text-sm text-muted-foreground"
          id=":r1:-form-item-description"
        >
          This will be used for logging in.
        </p>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
